src/server/tablet_snap.rs

```diff
@@ -101,67 +176,381 @@ fn io_type_from_raft_message(msg: &RaftMessage) -> Result<IoType> {
-        buffer.push(name.len() as u8);
-        buffer.extend_from_slice(name.as_bytes());
-        let mut f = File::open(&path)?;
-        let mut off = buffer.len();
-        loop {
-            unsafe {
-                buffer.set_len(SNAP_CHUNK_LEN);
+fn protocol_error(exp: &str, act: impl Debug) -> Error {
+    Error::Other(format!("protocol error: expect {exp}, but got {act:?}").into())
+}
+
+async fn cleanup_cache(
+    path: &Path,
+    stream: &mut (impl Stream<Item = Result<TabletSnapshotRequest>> + Unpin),
+    sink: &mut (impl Sink<(TabletSnapshotResponse, WriteFlags), Error = grpcio::Error> + Unpin),
+    limiter: &Limiter,
+) -> Result<(u64, Vec<String>)> {
+    let mut reused = 0;
+    let mut exists = HashMap::default();
+    for entry in fs::read_dir(path)? {
+        let entry = entry?;
+        let ft = entry.file_type()?;
+        if ft.is_dir() {
+            fs::remove_dir_all(entry.path())?;
+            continue;
+        }
+        if ft.is_file() {
+            let os_name = entry.file_name();
+            let name = os_name.to_str().unwrap();
+            if is_sst(name) {
+                // Collect length requires another IO, delay till we are sure
+                // it's probably be reused.
+                exists.insert(name.to_string(), entry.path());
+                continue;
             }
-            // it should break if readed len is zero or the buffer is full.
-            while off < SNAP_CHUNK_LEN {
-                let readed = f.read(&mut buffer[off..])?;
-                if readed == 0 {
-                    unsafe {
-                        buffer.set_len(off);
+        }
+        fs::remove_file(entry.path())?;
+    }
+    let mut missing = vec![];
+    loop {
+        let mut preview = match stream.next().await {
+            Some(Ok(mut req)) if req.has_preview() => req.take_preview(),
+            res => return Err(protocol_error("preview", res)),
+        };
+        let mut buffer = Vec::with_capacity(PREVIEW_CHUNK_LEN);
+        for meta in preview.take_meta().into_vec() {
+            if is_sst(&meta.file_name) && let Some(p) = exists.remove(&meta.file_name) {
+                let mut f = File::open(&p)?;
+                let exist_len = f.metadata()?.len();
+                if exist_len == meta.file_size {
+                    let head_len = meta.head_chunk.len();
+                    let trailing_len = meta.trailing_chunk.len();
+                    if head_len as u64 > meta.file_size || trailing_len as u64 > meta.file_size {
+                        return Err(Error::Other(format!("invalid chunk length {} {} {}", meta.file_size, head_len, trailing_len).into()));
+                    }
+                    // Usually the file is in page cache.
+                    read_to(&mut f, &mut buffer, head_len, limiter).await?;
+                    if buffer == meta.head_chunk {
+                        if meta.trailing_chunk.is_empty() {
+                            reused += meta.file_size;
```

Comment:
What does it mean if trailing_chunk is empty?

Reply:
It means the file is too small that `head_chunk` already contains all the data.